packages/graph/hash_graph/lib/interval-ops/src/interval.rs

```diff
@@ -193,84 +221,1191 @@ pub trait Interval<T>: Sized {
+        IntervalBounds::from_bounds(Bound::Included(lower), Bound::Excluded(upper))
+    }
+
+    fn excluded_included(lower: u32, upper: u32) -> IntervalBounds<u32> {
+        IntervalBounds::from_bounds(Bound::Excluded(lower), Bound::Included(upper))
+    }
+
+    fn excluded_excluded(lower: u32, upper: u32) -> IntervalBounds<u32> {
+        IntervalBounds::from_bounds(Bound::Excluded(lower), Bound::Excluded(upper))
+    }
+
+    #[test]
+    fn test_partially_overlapping() {
+        // Range A:      [-----]   |   [-----]
+        // Range B:        [-----] | [-----]
+        // intersection:   [---]   |   [---]
+        // union:        [-------] | [-------]
+        // merge:        [-------] | [-------]
+        // difference:   [-)       |       (-]
+        test(
+            included_included(0, 10),
+            included_included(5, 15),
+            [included_included(5, 10)],
+            [included_included(0, 15)],
+            [included_included(0, 15)],
+            [included_excluded(0, 5)],
+        );
+        test(
+            included_included(5, 15),
+            included_included(0, 10),
+            [included_included(5, 10)],
+            [included_included(0, 15)],
+            [included_included(0, 15)],
+            [excluded_included(10, 15)],
+        );
+
+        // Range A:      [-----)   |   [-----]
+        // Range B:        [-----] | [-----)
+        // intersection:   [---)   |   [---)
+        // union:        [-------] | [-------]
+        // merge:        [-------] | [-------]
+        // difference:   [-)       |       [-]
+        test(
+            included_excluded(0, 10),
+            included_included(5, 15),
+            [included_excluded(5, 10)],
+            [included_included(0, 15)],
+            [included_included(0, 15)],
+            [included_excluded(0, 5)],
+        );
+        test(
+            included_included(5, 15),
+            included_excluded(0, 10),
+            [included_excluded(5, 10)],
+            [included_included(0, 15)],
+            [included_included(0, 15)],
+            [included_included(10, 15)],
+        );
+
+        // Range A:      [-----]   |   [-----)
+        // Range B:        [-----) | [-----]
+        // intersection:   [---]   |   [---]
+        // union:        [-------) | [-------)
+        // merge:        [-------) | [-------)
+        // difference:   [-)       |       (-)
+        test(
+            included_included(0, 10),
+            included_excluded(5, 15),
+            [included_included(5, 10)],
+            [included_excluded(0, 15)],
+            [included_excluded(0, 15)],
+            [included_excluded(0, 5)],
+        );
+        test(
+            included_excluded(5, 15),
+            included_included(0, 10),
+            [included_included(5, 10)],
+            [included_excluded(0, 15)],
+            [included_excluded(0, 15)],
+            [excluded_excluded(10, 15)],
```

Comment:
How does this turn into a `excluded_excluded` when using `difference`? I assume it isn't wrong, but it's strange to see :)

Reply:
The upper bound of `lhs` is excluded and the upper bound of `rhs` is included. The upper bound of `lhs` is included in the difference, but the upper bound of `rhs` is not (as it's subtracted).

Reply:
You can check the result by running
```sql
SELECT numrange(5, 15, '[)') - numrange(0, 10, '[]');
```